src/persist-client/src/cli/inspect.rs

```diff
@@ -630,6 +643,204 @@ pub async fn unreferenced_blobs(args: &StateArgs) -> Result<impl serde::Serializ
+    struct BlobReport {
+        num_shards: usize,
+        num_blobs: usize,
+        missing_blobs: usize,
+        num_rows: usize,
+        num_bytes: usize,
+    }
+
+    /// Helper method that lists all shards in a Blob store.
+    async fn list_shards(
+        blob: Arc<dyn Blob + Send + Sync>,
+    ) -> Result<BTreeMap<ShardId, Vec<PartialBatchKey>>, anyhow::Error> {
+        let mut shards = BTreeMap::new();
+        blob.list_keys_and_metadata(&BlobKeyPrefix::All.to_string(), &mut |metadata| {
+            match BlobKey::parse_ids(metadata.key) {
+                Ok((shard, PartialBlobKey::Batch(writer_key, part_id))) => {
+                    let partial_keys = shards.entry(shard).or_insert_with(Vec::new);
+                    partial_keys.push(PartialBatchKey::new(&writer_key, &part_id));
+                }
+                Err(err) => {
+                    eprintln!("error parsing blob: {}", err);
+                }
+                _ => (),
+            }
+        })
+        .await?;
+
+        Ok(shards)
+    }
+
+    // Open a handle to S3.
+    let blob = make_blob(&cfg, blob_uri, NO_COMMIT, Arc::clone(&metrics))
+        .await
+        .context("making blob")?;
+    tracing::info!(?blob_uri, "opened blob");
+
+    // List all shards.
+    let shards = mz_ore::retry::Retry::default()
+        .max_tries(5)
+        .retry_async(|_| {
+            let blob = Arc::clone(&blob);
+            async move { list_shards(Arc::clone(&blob)).await }
+        })
+        .await?;
+
+    let num_parts: usize = shards.values().map(|parts| parts.len()).sum();
+    tracing::info!(
+        ?blob_uri,
+        num_shards = shards.len(),
+        num_parts,
+        "listed blob"
+    );
+
+    let mut report = BlobReport::default();
+    for (shard_id, partial_keys) in shards {
+        // Drives all of the GET futures concurrently.
+        //
+        // Note(parkmycar): We should limit the number of blobs we fetch in parallel to reduce
+        // memory usage.
+        let mut blob_stream: FuturesUnordered<_> = partial_keys
+            .iter()
+            .map(|partial_key| {
+                let blob = Arc::clone(&blob);
+                let key = partial_key.complete(&shard_id);
+                mz_ore::retry::Retry::default().retry_async(move |_| {
+                    let blob = Arc::clone(&blob);
+                    let key = key.clone();
+                    async move { blob.get(&key).await }
+                })
+            })
+            .collect();
+
+        // Updates the 'persist_use_arrow_rs_library' feature flag.
+        let update_feature_flag = |val: &str| {
+            let mut updates = mz_dyncfg::ConfigUpdates::default();
+            let val = mz_dyncfg::ConfigVal::String(val.to_string());
+            updates.add_dynamic("persist_use_arrow_rs_library", val);
+            updates.apply(metrics.columnar.cfg());
+        };
+
+        while let Some(blob_result) = blob_stream.next().await {
```

Comment:
This `FuturesUnordered` plus compute-heavy work on the result of `.next().await` is something I consider an anti-pattern. While the compute-heavy work is happening in between `.next().await` calls, none of the futures are being driven forward. As persist has now learned several times, not being prompt about calling `.poll` on an IO future after it's waker is pinged creates all kinds of sadness.

(My personal rule of thumb is now to _never_ use `FuturesUnordered` unless for `.collect`)

Anyway, the fix is easy: use tasks! Each these are big IO fetches, so no worries about overhead.

Reply:
This totally makes sense, I forgot about how the futures won't be driven forward in-between calls to `.next().await`. Changed this to use tasks!